### PR TITLE
CEL expressions approver design

### PR DESCRIPTION
**Pull request by erikgb**
This PR contains an initial design proposal allowing policy authors to use CEL expressions to express policy rules. If this is ever implemented, it should address https://github.com/cert-manager/approver-policy/issues/62.

Also includes an initial design template copied from [cert-manager](https://github.com/cert-manager/cert-manager/blob/master/design/template.md).

/cc @inteon @SgtCoDFish

### CODE REVIEW

File: design/20230726-cel-policy.md
Comment (erikgb) [inline design/20230726-cel-policy.md:204]:
```suggestion
      validations:
        - rule: "self.endsWith(cr.namespace + '.apps.my-cluster.com')"
          message: "DNSName should be in cluster apps domain"
```

After a bit of thinking, I think we should align our design closer to upstream [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules).

I'll update the suggested design now.

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:103]:
```suggestion
Allowed string fields accept wildcards `"*"` within its value(s).
```
? Or `*` ...

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:56]: Is there a reason to point to that instead of https://kubernetes.io/docs/reference/using-api/cel/?

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:124]:
```suggestion
"trip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:133]: It'd be nice if `SVID` and `SPIFFE ID` were links to things.

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:56]: We could link to both. The Kubernetes docs give a good overview, but if you're interested in the details, I think you need to consult the detailed spec.

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md]: Thanks @jsoref! Updated with links.

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:165]: this line is awkward. https://en.wikipedia.org/wiki/Garden-path_sentence

In general, I'd discourage the use of abbreviations (`'`), 
```suggestion
requested features exists, and we propose to base the design on this plugin.
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:356]:
```suggestion
  This could increase the maintenance burden and lead to subtle bugs. Breaking the API could be an option to mitigate this.
```

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:56]: Good point, thanks! I will change the link in my next round on this. Also thanks for the rewording suggestions! 👍

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md]: Modified the proposal to reuse the Kubernetes CEL validations.

**Comment by erikgb** [comment]
I have now updated the proposal and moved the previous proposal into the alternatives. The new proposal is now significantly closer to the upstream CEL validations rules API. Please take another look. I would love to see some progress on this design proposal so we eventually can move over to implementing it. 😸 

/cc @jsoref @inteon @SgtCoDFish @wallrj

**Review by SgtCoDFish — COMMENTED** [review]
/lgtm
/approve
/hold

(Holding so others can chime in)

First of all, this is well-written and makes sense to me. I understand the rationale and thank you for your patience with this!

I think the idea here is solid. It adds a lot of flexibility and power and the tradeoff in complexity is worth it.

A big part of why I'm really relaxed about this is that approver-policy isn't super-widely used and we're still at v1alpha1, which to me means we should be more relaxed about making changes and trying things out.

I think we should proceed with this on the basis that we _should_ be experimenting with this. I'd also support making these changes in v1alpha2, leaving v1alpha1 as what we have today, which in my mind should make this even easier to approve. But I don't think that's required - I'd be happy to add this to v1alpha1.

As for the alternatives: I instinctively prefer having expressions be tied to fields rather than operating on the whole CSR (i.e. I prerfer the proposal to that alternative). I think the simpler the CEL the better.

**Review by inteon — APPROVED** [review]
+1

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:5]:
```suggestion
# CEL validation rules to approve or deny a CertificateRequest
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:55]:
```suggestion
enforce CSR attributes that must be a function of the CertificateRequest's namespace (or other CertificateRequest fields).
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:75]:
```suggestion
- allow policy authors to express `allowed` CSR attribute value(s) to be a function of CertificateRequest's fields using CEL expressions
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:103]: possibly?
```suggestion
`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the requested attribute.
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:108]:
```suggestion
wildcard support, the main improvement would be support for policies that specify an allowed attribute value that depends
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:114]:
```suggestion
The main downside of using a plugin for this, in addition to increased maintenance/complexity,
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:117]:
```suggestion
Which means that approver rules are somehow duplicated in the `allowed` block and in the plugin configuration but
this should be avoided as maintaining things in multiple places is error prone.
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:333]:
```suggestion
### CEL expressions with semantics similar to values
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:335]:
```suggestion
_Note: This was the original proposal, after some consideration it was considered inferior to the current proposal,
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:384]:
```suggestion
denied. An example policy with this alternative API might look like (example):
```

File: design/20230726-cel-policy.md
Comment (inteon) [inline design/20230726-cel-policy.md:357]: Maybe, it would be useful if we could also express that eg. you can have at most 3 dnsNames in your cert?

File: design/20230726-cel-policy.md
Comment (inteon) [inline design/20230726-cel-policy.md:205]: How could we limit the number of dnsNames? Could we eg. have a "selfs" parameter?

File: design/20230726-cel-policy.md
Comment (inteon) [inline design/20230726-cel-policy.md:190]: I would start with a `cr` value that only contains the basic fields of the CertifiicateRequest resource (name, namespace, annotations, labels, uuid, ....). Once that works, we could also add other properties (like x509 stuff).

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:418]: I have now looked at the source code, and approver-policy only evaluates a policy against the CSR attributes that approver-policy supports. So you are correct that it might be considered too permissive in this sense. I suggest creating an issue for this if it is something that we should attempt to fix. Changing it would be breaking tho. 😉

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:205]: This is an interesting and important comment! Strictly speaking, many CSR attributes are multi-valued - so it could make sense to make `self` refer to the list of attribute values in the CSR - which will allow for the validation you are requesting here. But I suspect this will make the feature less user-friendly. The validation example this review comment refers to would then have to be expressed as `self.all(dnsName, dnsName.size() =< 64)` - which is not too bad IMO.

But this could easily lead to user errors like `self.size() =< 64` would now limit the number of dnsNames in the CSR to 63. 😉 Which is probably not what the user had in mind.

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:190]: Agreed, I think we should start even simpler in the initial implementation and provide just CR `namespace` and `name`. I will add this initial MVP to the proposed design.

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:205]: IMO the most user-friendly is to keep the validations under `allowed`  as per-item validations. This is how the `value`/`values` fields in the current API work. More complex CEL validations, typically cross-cutting concerns could be introduced later outside of the `allowed` API structure. I am updating the proposal with this clarification now.

File: design/20230726-cel-policy.md
Comment (erikgb) [reply design/20230726-cel-policy.md:197]: Updated, PTAL!

**Review by SgtCoDFish — APPROVED** [review]
/lgtm
/approve

Fundamentally, I think we should be experimenting with approver-policy. Now's the time for it!

I'd be much more conservative about this if it were for cert-manager itself... but it's not. I think this is worth a go. It makes sense. It's well written. Thank you for this Erik! 👍

**Comment by erikgb** [comment]
I have initiated the work on implementing this new feature: https://github.com/cert-manager/approver-policy/pull/277
Please take a look if you have time.

**Comment by erikgb** [comment]
After working on improving the API docs and understanding more of the current API behavior, I have updated the proposed design (and examples with the following):
>Note: To ensure we don't break behaviors in the existing API version, any validations can never allow a request otherwise denied by the `values`/`value` and `required` fields.

I think that is our safest option. The next API version could make this feature even more user-friendly. But I think it will be acceptable as proposed now.

I am almost ready to move over to the implementation of this, so I would prefer a more "formal" approval of this design proposal by merging this PR. 😉 PTAL!

**Comment by hawksight** [comment]
Just wanted to add that the design I read just over 2 weeks ago looked great. It covered a few things I had not thought of and there was nothing I disagreed with. Now work has started, I will attempt to look over for understanding.

The only minor thing is that there are a few empty sections in the design doc, might it be worth either a quick comment of `N/A` if the section isn't applicable.

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:105]:
```suggestion
Allowed string fields accept wildcards `"*"` within their values.
```
I think...

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:107]:
```suggestion
This proposal suggests adding support for specifying allowed string fields using CEL expressions, similar to the existing
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:182]:
```suggestion
validations pass for **ALL* attribute values, the attribute will be allowed.
```

File: design/20230726-cel-policy.md
Comment (jsoref) [inline design/20230726-cel-policy.md:188]:
```suggestion
* if `required` is set to `true` in policy and attribute value empty in request, the request will be denied
  by policy without running any validations.
* if `required` is set to `true` in policy, the policy will need specify `values`/`value` or `validations`.
```

**Pull request merged**
